db/db_iter.cc

```diff
@@ -1060,10 +1077,20 @@ bool DBIter::TooManyInternalKeysSkipped(bool increment) {
   return false;
 }
 
-bool DBIter::IsVisible(SequenceNumber sequence) {
+bool DBIter::IsVisible(SequenceNumber sequence, const Slice& ts) {
+  int cmp_ts = timestamp_ub_ != nullptr
+                   ? user_comparator_.CompareTimestamp(ts, *timestamp_ub_)
+                   : 0;
+  if (cmp_ts < 0) {
+    return false;
+  } else if (cmp_ts > 0) {
+    assert(sequence <= sequence_);
```

Comment:
Can you explain why this assertion?

Reply:
`cmp_ts>0` -> `key.ts > ts_ub` -> key.ts is before ts_ub. We currently assume timestamp order is the same as sequence number order. Thus, the key.seq is smaller than (before) snapshot.seq.

Reply:
This is a good point. Since `sequence_` can also come from user, it is not a good practice to assert on its value. I will remove the assertion. It will be tricky if user specifies both timestamp and sequence number.

Reply:
I had to change this to:
```c
if (sequence <= sequence_) {
  return true;
}
return false;
```

otherwise this line failed the assertion: https://github.com/BradBarnich/rocksdb/commit/00d7c3107d81920d38f879b9a5e9848bce7159df#diff-4b40bd5d318908a16d60d610ff1974eaR3311

(which is a copy of IterWithSnapshot with user timestamps, but it relies on reverse iteration support.)

There might be a bug in my reverse iteration support though

Reply:
Do you mean you have to remove the entire comparison of timestamp from the `IsVisible` method? I believe the `IsVisible` should consider timestamp too. I am not familiar with reverse iterator yet, but I think a fix may be needed elsewhere.

Reply:
No, just replaced the assertion line with this if statement.